### PR TITLE
[DataLoader] Apply sharding settings in dist when num_workers is 0

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #78950

Tested locally with 
```python
import torch
from torch.utils.data import DataLoader

from torchdata.datapipes.iter import IterableWrapper

import os
import sys
import tempfile
import torch
import torch.distributed as dist
import torch.nn as nn
import torch.optim as optim
import torch.multiprocessing as mp

def cleanup():
    dist.destroy_process_group()

def setup(rank, world_size):
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '12355'
    dist.init_process_group("gloo", rank=rank, world_size=world_size)

def my_function(worker_id):
    print('worker', worker_id)

setup(int(os.environ['RANK']), int(os.environ['WORLD_SIZE']))
dp = IterableWrapper(range(100)).shuffle().sharding_filter()
dl = DataLoader(dp, num_workers = 1, worker_init_fn = my_function)
print(sorted(list(dl)))

cleanup()
```

### CODE REVIEW

**Review by ejguan — APPROVED** [review]
Thank you. LGTM

**Review by NivekT — APPROVED** [review]
LGTM! It is possible to make the code more concise by editing `_sharding_worker_init_fn` but I think it is more readable as it is.

nit: Add commetns to `_apply_distributed_sharding` and `_sharding_worker_init_fn` so future readers will know why these functions need to exist.

**Comment by VitalyFedyunin** [comment]
These two functions will require some refactoring (following the seed exchange example we can also exchange the number of processes per rank). I will try to merge them into 1.12 and follow-up with update after that to drop 'equal amount of processes' restriction.
